AEP/aep-00004.md

```diff
@@ -0,0 +1,122 @@
+# AEP 4: ArmoniK Load Balancer
+
+|                   |ArmoniK Enhancement Proposal|
+---:                |:---
+**AEP**             | 4
+**Title**           | ArmoniK Load Balancer
+**Author**          | Florian Lemaitre <<flemaitre@aneo.fr>>
+**Status**          | Draft
+**Type**            | Standard
+**Creation Date**   | 2024-04-12
+
+# Abstract
+
+This AEP proposes an ArmoniK native Load Balancer, which can be used to scale ArmoniK clusters.
+
+# Motivation
+
+ArmoniK is built around a centralized database, queue and object storage, and that makes it hard to scale. While it is fairly easy to have a scalable (distributed) object, it is much more complex to setup a scalable database or queue.
+
+# Rationale
+
+A simple alternative is to have multiple isolated ArmoniK clusters, with their own database, queue, and object storage. To make such a design transparent and easy to work with, an ArmoniK Load Balancer would be needed.
+
+Such a Load Balancer would understand the ArmoniK concepts and forwards user requests based on ArmoniK data like the session ID.
+
+# Specification
+
+## solution overview
+
+An ArmoniK Load Balancer would be an external service that would serve ArmoniK API in gRPC and redirect them to a cluster.
+Upon Session creation, a single cluster is assigned to the session, chosen by the implementation. Afterwards, all requests targeting the session will be redirected to its assigned cluster.
```

Comment:
Shouldn't the user be able to specify the cluster targeted ?
How does this cope with the choice of partitions ?

Reply:
Meta-control plane could look into task options or session options to get indications to make the choice

Reply:
@wkirschenmann This is explained later on, with the Database implementation when I talk about smart decisions. I could add a sentence there to emphasize the possibility to choose according to the task options.

Regarding the partitions, if the load balancer cannot find a cluster with all your partitions constraints, it will raise an error as explained in the partition section. Forcing a cluster using metadata does not change that, and just reduce the number of clusters considered for selection.

Reply:
If I understant it correctly, that means that tasks from a session which start on prem cannot be done on the cloud?

Reply:
Yes, that's what is explained.